feat(files_sharing): Add `types` filter to share API

nfebe opened the pull request:
# Description

Add share filtering to Share API

nfebe committed:
feat(files_sharing): Add `types` filter to `getShares`

Improve the OCS files sharing API to include a share type filter in `getShares`

Signed-off-by: nfebe <fenn25.fn@gmail.com>

nfebe committed:
feat(sharing_tab): Show remote shares in external shares section

Signed-off-by: nfebe <fenn25.fn@gmail.com>

susnux reviewed (commented):
Not sure we need two requests to the backend, we also can just filter on the frontend

provokateurin reviewed (changes requested):
Agreed, the filtering should be done on the frontend.
Also this breaks other API users because you introduce a new field without a default value that is the same as the current behavior.

susnux commented:
> Generally, this increases performance, especially for instances/users with a sizable number of shares, Splitting and parallelizing the request has clear benefits and some drawbacks and so does frontend filtering.

I am not sure here.
1. the sidebar loading will increase as the second request itself will add a delay (network)
2. You need to setup the whole enviroment (filesystem / shares) two times as we never ask for only one of both, so this basically duplicates the request time. Meaning if in worst case there is no other php process available this now will take twice the time as before and even if handled sequentially it will double the server load.

So I do not see any benefit here as the shares are requested the same way, there is no shortcut taken while fetching the shares, only formatting them is skipped. But this means now we have two fetch **all** shares 2 times iterate overthem and filter them.

nfebe committed:
feat: Only getShares for provided types

With the types filter in place (1e7f53e9bb791c151b5aa637974f6689386ebd33) we should avoid,
fetching all shares.

Preventing potential N+1s

Signed-off-by: nfebe <fenn25.fn@gmail.com>

nfebe reviewed (commented):
I am aware of the missing openapi update, had some issues with `composer exec generate-spec`.

While the main reviews are being cleared that would be handled.

nfebe commented:
/compile

provokateurin commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -1065,6 +1092,11 @@ private function getFormattedShares(
 			}
 
+			// Filter out share types not requested in non-empty $types array
+			if (!empty($types) && !in_array($share->getShareType(), $types)) {
+				continue;
+			}
```
Only doing the filtering after already querying all of them is super inefficient and can simply be done on the frontend side as well.
I really don't see how this is an improvement, especially given that we could reduce the frontend requests to one and just filter/split there.

susnux replied:
yes either filter BEFORE querying and only query whats needed to speed up or get all and filter in the frontend.
Especially if you need to query all in all cases (and just show them in different sections) then one single request is faster because you do not have the overhead of setting up the filesystem on the server two times and you save the second network round trip.

nfebe replied:
Moved quick fix to : https://github.com/nextcloud/server/pull/52423

nfebe commented:
I've moved the fix to: https://github.com/nextcloud/server/pull/52423.
I'll treat this as a non-backportable addition, which also gives us more time to address any related performance issues without blocking the original fix.
